prow/cmd/checkconfig/main.go

```diff
@@ -1104,3 +1113,120 @@ func validateCluster(cfg *config.Config) error {
+		isRepoConfig := nestingDepth == 2
+		isNestedTooDeep := nestingDepth > 2
+
+		// Might contain anything or not even be a Prow config.
+		if isGlobalConfig {
+			return nil
+		}
+
+		if isNestedTooDeep {
+			errs = append(errs, fmt.Errorf("config %s is at an invalid location. All configs must be below %s. If they are org-specific, they must be in a folder named like the org. If they are repo-specific, they must be in a folder named like the repo below a folder named like the org.", path, root))
+			return nil
+		}
+
+		rawCfg, err := fs.ReadFile(filesystem, path)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("failed to read %s: %w", path, err))
+			return nil
+		}
+		var prowCfg config.ProwConfig
+		if err := yaml.Unmarshal(rawCfg, &prowCfg); err != nil {
+			errs = append(errs, fmt.Errorf("failed to deserialize config at %s: %w", path, err))
+			return nil
+		}
+
+		isGlobal, targetedOrgs, targetedRepos := prowCfg.HasConfigFor()
+		if isOrgConfig {
+			expectedTargetOrg := pathElements[0]
+			if !isGlobal && len(targetedOrgs) == 1 && targetedOrgs.Has(expectedTargetOrg) && len(targetedRepos) == 0 {
+				return nil
+			}
+			errMsg := fmt.Sprintf("config %s is invalid: Must contain only config for org %s, but", path, expectedTargetOrg)
+			var needsAnd bool
+			if isGlobal {
+				errMsg += " contains global config"
+				needsAnd = true
+			}
+			for _, org := range targetedOrgs.Delete(expectedTargetOrg).List() {
+				errMsg += prefixWithAndIfNeeded(fmt.Sprintf(" contains config for org %s", org), needsAnd)
+				needsAnd = true
+			}
+			for _, repo := range targetedRepos.List() {
+				errMsg += prefixWithAndIfNeeded(fmt.Sprintf(" contains config for repo %s", repo), needsAnd)
+				needsAnd = true
+			}
+			errs = append(errs, errors.New(errMsg))
+			return nil
+		}
+
+		if isRepoConfig {
+			expectedTargetRepo := pathElements[0] + "/" + pathElements[1]
+			if !isGlobal && len(targetedOrgs) == 0 && len(targetedRepos) == 1 && targetedRepos.Has(expectedTargetRepo) {
+				return nil
+			}
+
+			errMsg := fmt.Sprintf("config %s is invalid: Must only contain config for repo %s, but", path, expectedTargetRepo)
+			var needsAnd bool
+			if isGlobal {
+				errMsg += " contains global config"
+				needsAnd = true
+			}
+			for _, org := range targetedOrgs.List() {
+				errMsg += prefixWithAndIfNeeded(fmt.Sprintf(" contains config for org %s", org), needsAnd)
+				needsAnd = true
+			}
+			for _, repo := range targetedRepos.Delete(expectedTargetRepo).List() {
+				errMsg += prefixWithAndIfNeeded(fmt.Sprintf(" contains config for repo %s", repo), needsAnd)
+				needsAnd = true
+			}
+			errs = append(errs, errors.New(errMsg))
+			return nil
+		}
+
+		// We should have left the function earlier. Error out so bugs in this code can not be abused.
+		return fmt.Errorf("BUG: You should never see this. Path: %s, isNestedTooDeep: %t, isGlobal: %t, targetedOrgs: %v, targetedRepos: %v", path, isNestedTooDeep, isGlobal, targetedOrgs, targetedRepos)
```

Comment:
nit: you could do a `switch` on nesting depth to be explicitly exhaustive on these branches if you wanted to

Reply:
Done